### PR TITLE
Generalization of Quantum Metric Learning Classifiers Demo Integration

### PR DESCRIPTION
**Pull request by Rlag1998**
------------------------------------------------------------------------------------------------------------

**Title:**
Generalization of Quantum Metric Learning Classifiers

**Summary:**
These demos are a fork of the previously discontinued Embeddings & Metric Learning demo authored by Maria Schuld and Aroosa Ijaz in 2020. These new demos use the UCI ML Breast Cancer (Diagnostic) Dataset and the ImageNet ants/bees image dataset to assess the generalization limits and performance of quantum metric learning. Schuld and Ijaz's original code was adapted in numerous ways to attempt to produce good test set results for both datasets. The ants/bees dataset, which had a high number of initial features per sample, did not lead to good generalization. Models generalized best for test data when a smaller number of features per sample was used (as seen with the breast cancer dataset), particularly after feature reduction through principal component analysis. Ultimately, these demos illustrate that quantum metric learning can lead to accurate test set classification given a suitable dataset and appropriate data preparation.

**Relevant references:**
Seth Lloyd, Maria Schuld, Aroosa Ijaz, Josh Izaac, Nathan Killoran: "Quantum embeddings for machine learning"
arXiv preprint arXiv:2001.03622.

Andrea Mari, Thomas R. Bromley, Josh Izaac, Maria Schuld, Nathan Killoran: "Transfer learning
in hybrid classical-quantum neural networks" arXiv preprint arXiv:1912.08278

"Generalization Performance of Quantum Metric Learning Classifiers" by Jonathan Kim and Stefan Bekiranov submitted to Biomolecules (currently in review stage).

An additional README file for the demos can be found [here](https://github.com/Rlag1998/QML_Generalization)

**Possible Drawbacks:**
*tutorial_embedding_generalization_antbee_original.py* and *tutorial_embedding_generalization_antbee_pca.py* are somewhat redundant. The most important demo here is *tutorial_embedding_generalization.py*, which uses the breast cancer dataset successfully to show good generalization performance.

**Related GitHub Issues:**
This demo does not work on versions of pennylane after 0.18.0, seemingly due to parameter recognition errors. 

Listed below are the requirements:

pennylane==0.18.0
numpy==1.22.3
scikit-learn==1.0.2
scipy==1.8.0
pandas==1.4.1
matplotlib==3.5.2
torch==1.11.0
torchvision==0.12.0

There are currently three demos we are proposing. The main one is tutorial_embedding_generalization which classifies breast cancer data. The other two are more supplementary in nature, as they demonstrate that having too many features per sample (such as when using the ants/bees image dataset) results in bad generalization.

### CODE REVIEW

**Comment by KetpuntoG** [comment]
Hello again @Rlag1998 ! :)  Thank you very much for doing the PR! I have been working on the three demos. As you may know, this is a demo that we had some time ago and we decided to remove from the site as the results were not quite good. I have seen that in the initial dataset, (bees and ants) you have not been able to improve the previous results. Although the results in the other demo are better, the dataset is much simpler. In order to replace the demo with ours, we would like to see an improvement in the first dataset proposed. If not, we could help you to add the cancer demo to the [community section](https://pennylane.ai/qml/demos_community.html) instead, a good place also to share the work 😃

**Comment by Rlag1998** [comment]
Hello again @KetpuntoG!  You're welcome and thank you for your feedback and help with getting our demos up on pennylane!  Importantly, in our submitted paper and these demos, we show that quantum metric learning satisfies the traditional bias-variance constraints of statistical learning theory. Because the ant-bees dataset violated these rules (many more fitting parameters than samples), it overfit the data, leading to poor performance on test data. Fixing this would require a new method going beyond quantum metric learning. The breast cancer data satisfies traditional bias-variance constraints (fewer fitting parameters than samples). We believe a QML demo which makes the above points clear on pennylane would be valuable to the quantum machine learning community. We also do want to reiterate an important technical point: the demo does not work on versions of pennylane after 0.18.0.

Listed below are the requirements:


pennylane==0.18.0
numpy==1.22.3
scikit-learn==1.0.2
scipy==1.8.0
pandas==1.4.1
matplotlib==3.5.2
torch==1.11.0
torchvision==0.12.0

Thanks!

**Comment by KetpuntoG** [comment]
I've been thinking, why doesn't it work for you in the new versions. It's probably because you are not training the parameters. In pennylane [version `0.19`](https://pennylane.ai/blog/2021/11/pennylane-v019-released/) this change was introduced:

Captura de Pantalla 2022-10-18 a las 11 39 09

Make sure that the parameters are trainable and the dataset elements are not.

Okay, so let's do the following. Let's work with the cancer demo  and see if we can adapt it to our standards.👌 The other two will not be included so they could be removed from the PR with their respective images. 

I would ask you to avoid introducing the dataset by importing it from a file. Probably this dataset is in some other library that you can use and you can import it from there. Since the strength of the paper from what you say is the study of the number of parameters and the dataset, I think it is something that at a theoretical level would be important to find in the demo 😄

**Comment by Rlag1998** [comment]
I've now removed the other two demos (the ants/bees demos )from the PR.

At the moment, although this was not the case with the ants/bees data (which were imported from image files), the breast cancer data is being imported from sklearn:
image
I'm guessing this should be fine? 😄 

We have results tables from our paper demonstrating the effects PCA/varying the number of input features to the quantum feature map has on training cost, test cost, precision, recall and F1 score - perhaps we can add one of these results tables to the demo?

Thanks!

**Comment by KetpuntoG** [comment]
Could you also delete the ants/bees images? 
 In this case, it would be interesting to put the data processing code and explain it as it seems to be something important in your development. We usually avoid calling other `.py` documents. 

At the beginning of the demo you comment: "After all necessary data pre-processing steps". This is what needs to be explained, not leaving steps that the reader does not know what is going on :)

**Comment by Rlag1998** [comment]
I've now removed the images too. The data preprocessing code actually varies depending on whether PCA is used or not, so there are two different preprocessing python files. However, I could perhaps add the PCA version, as it allows for most variation of the number of input parameters?

**Comment by KetpuntoG** [comment]
> I've now removed the images too. The data preprocessing code actually varies depending on whether PCA is used or not, so there are two different preprocessing python files. However, I could perhaps add the PCA version, as it allows for most variation of the number of input parameters?

Sounds good to me 😄

**Comment by Rlag1998** [comment]
I have now added the preparation code to the start of the demo and have added a table of results at the end summarising the effects of using different numbers of principal components.

I have tried numerous different things to get the parameter training to work on the latest version of pennylane (e.g., applying requires_grad = True to both init_pars_quantum and init_pars_classical), but with no success - I think it's struggling with the fact that it's training an array made up of two other arrays (init_pars). I don't suppose it would be possible to set this specific demo to run on 0.18.0 (where we can then add a clear note in the demo to say that ver. 0.18.0 is required)?

**Comment by KetpuntoG** [comment]
> I have now added the preparation code to the start of the demo and have added a table of results at the end summarising the effects of using different numbers of principal components.
> 
> I have tried numerous different things to get the parameter training to work on the latest version of pennylane (e.g., applying requires_grad = True to both init_pars_quantum and init_pars_classical), but with no success - I think it's struggling with the fact that it's training an array made up of two other arrays (init_pars). I don't suppose it would be possible to set this specific demo to run on pennylane version 0.18.0 (where we can then add a clear note in the demo to say that ver. 0.18.0 is required)?

Our demos should always be updated to the latest version of PennyLane, maybe you have found some kind of bug.
I will check your code to see if I can find the problem

**Comment by KetpuntoG** [comment]
Hey @Rlag1998 , I've been discussing it with the team and we decided that this demo would fit better in the community section 🙂 One of the advantages of that is that you will be able to use the versions of all the libraries that you want. For that it would be interesting that you add a `requirements.txt`  to the repository. Here you have [the instructions](https://pennylane.ai/qml/demos_submission.html) on how to do that. When you have created the issue, you can tag me and I will take care of adding it on the web 💪

**Comment by Rlag1998** [comment]
After trying a few more things with the code, it definitely looks like it's not recognizing init_pars as being trainable, even though its components can be assigned requires_grad = True. I cannot assign init_pars itself the requires_grad = True attribute, as it is a list. I guess the thing is that in 0.18.0, 'everything' was trainable by default, but from 0.19.0, everything has to be set to be trainable manually and it seems there's not yet a way to set an array of arrays to be trainable in the new versions.

When it comes to adding a community demo, do you think it's still best to have only the breast cancer version, or do you reckon it can be more flexible and include the ants & bees versions too? In other words, should I create an issue with the current version of the demo (which includes the added data prep code at the start) or the older version with more data preparation files + ants/bees training files?

Thanks!

**Comment by KetpuntoG** [comment]
> After trying a few more things with the code, it definitely looks like it's not recognizing init_pars as being trainable, even though its components can be assigned requires_grad = True. I cannot assign init_pars itself the requires_grad = True attribute, as it is a list. I guess the thing is that in 0.18.0, 'everything' was trainable by default, but from 0.19.0, everything has to be set to be trainable manually and it seems there's not yet a way to set an array of arrays to be trainable in the new versions.
> 
> When it comes to adding a community demo, do you think it's still best to have only the breast cancer version, or do you reckon it can be more flexible and include the ants & bees versions too? In other words, should I create an issue with the current version of the demo (which includes the added data prep code at the start) or the older version with more/separate data preparation files + ants/bees training files?
> 
> Thanks!

The community demo will give you much more flexibility. You could upload all 3 😄
Only one link will appear on the web page. But it could be pointing to your repository directly. You could make a `readme.md` explaining that you have 3 demos, saying what you are going to find in each one and there you put the 3 links.

**Comment by Rlag1998** [comment]
I have now submitted an [issue](https://github.com/PennyLaneAI/qml/issues/615).

The issue refers to [this github repository](https://github.com/Rlag1998/QML_Generalization), where the readme.md and requirements.txt file are included along with the three demos 😄 

Thanks!

**Comment by KetpuntoG** [comment]
Hey @Rlag1998 ! Everything is ready and this Thursday we will publish the demo 😄
If you have a twitter account we could mention you on the tweet

**Comment by Rlag1998** [comment]
Hello @KetpuntoG, thank you for getting it all ready! 😄 

We don't really have Twitter accounts, but I've just made a fresh one: @Jonathan_Kim_98

Thanks!

**Comment by KetpuntoG** [comment]
I close the PR, again, thank you for your work!
